File: c++/All O`one Data Structure.cpp
#include <iostream>
#include <unordered_map>
#include <unordered_set>
#include <string>
#include <list>

class AllOne {
private:
    struct Bucket {
        int count;
        std::unordered_set<std::string> keys;
    };

    std::list<Bucket> buckets; 
    std::unordered_map<std::string, int> keyCount;  
    std::unordered_map<int, std::list<Bucket>::iterator> countBucket; 

    std::list<Bucket>::iterator addBucketAfter(std::list<Bucket>::iterator it, int count) {
        auto newBucket = buckets.insert(std::next(it), {count, {}});
        countBucket[count] = newBucket;  
        return newBucket;
    }

    void removeBucketIfEmpty(std::list<Bucket>::iterator it) {
        if (it->keys.empty()) {
            countBucket.erase(it->count);  
            buckets.erase(it);  
        }
    }

public:
    AllOne() {}

    void inc(std::string key) {
        int currentCount = keyCount[key]; 
        keyCount[key] = currentCount + 1;

        std::list<Bucket>::iterator currentBucket;

        if (currentCount > 0) {
            currentBucket = countBucket[currentCount];
            currentBucket->keys.erase(key); 
            removeBucketIfEmpty(currentBucket);  
        }

        int newCount = currentCount + 1; 
    
        if (countBucket.find(newCount) != countBucket.end()) {
            currentBucket = countBucket[newCount];  
        } else {
            currentBucket = (currentCount == 0) ? buckets.begin() : countBucket[currentCount];
            currentBucket = addBucketAfter(currentBucket == buckets.end() ? buckets.begin() : currentBucket, newCount);
        }
        currentBucket->keys.insert(key);  
    }

    void dec(std::string key) {
        int currentCount = keyCount[key];
        if (currentCount == 0) return; 

        keyCount[key] = currentCount - 1;  
        std::list<Bucket>::iterator currentBucket = countBucket[currentCount];
        currentBucket->keys.erase(key);  
        removeBucketIfEmpty(currentBucket); 

        if (currentCount == 1) {
            keyCount.erase(key);  
            return;
        }

        int newCount = currentCount - 1;

        if (countBucket.find(newCount) != countBucket.end()) {
            currentBucket = countBucket[newCount];  
        } else {
            if (currentBucket != buckets.begin()) {
                currentBucket = addBucketAfter(std::prev(currentBucket), newCount);  
            } else {
                currentBucket = buckets.insert(buckets.begin(), {newCount, {}});  
                countBucket[newCount] = currentBucket;
            }
        }
        currentBucket->keys.insert(key); 
    }

    std::string getMaxKey() {
        if (buckets.empty()) return ""; 
        return *(buckets.back().keys.begin());  
    }

    std::string getMinKey() {
        if (buckets.empty()) return "";  
        return *(buckets.front().keys.begin());  
    }
};
// Test case       
int main() {
    AllOne* obj = new AllOne();

    obj->inc("apple");
    obj->inc("banana");
    obj->inc("apple");
    obj->inc("orange");
    obj->inc("banana");
    obj->inc("banana");
    obj->inc("apple");

    std::cout << "Max Key: " << obj->getMaxKey() << std::endl; 
    std::cout << "Min Key: " << obj->getMinKey() << std::endl;  

    obj->dec("banana");
    std::cout << "Max Key after dec: " << obj->getMaxKey() << std::endl;  
    std::cout << "Min Key after dec: " << obj->getMinKey() << std::endl;  

    obj->dec("banana");
    obj->dec("banana");
    std::cout << "Max Key after more dec: " << obj->getMaxKey() << std::endl;  
    std::cout << "Min Key after more dec: " << obj->getMinKey() << std::endl; 
    
    delete obj;
    return 0;
}

File: python/All O`one Data Structure.py
class AllOne:
    def __init__(self):
        self.myDict = {}  
        self.countDict = {}  
        self.minVal = None
        self.maxVal = None

    def _updateMinMax(self):
        if not self.myDict:
            self.minVal = self.maxVal = None
        else:
            self.minVal = min(self.myDict.values())
            self.maxVal = max(self.myDict.values())

    def inc(self, key: str) -> None:
        if key in self.myDict:
            oldCount = self.myDict[key]
            newCount = oldCount + 1
            self.myDict[key] = newCount
            self.countDict[oldCount].remove(key)
            if not self.countDict[oldCount]:
                del self.countDict[oldCount]
        else:
            newCount = 1
            self.myDict[key] = newCount

        if newCount not in self.countDict:
            self.countDict[newCount] = set()
        self.countDict[newCount].add(key)

        self._updateMinMax()

    def dec(self, key: str) -> None:
        if key not in self.myDict:
            return

        oldCount = self.myDict[key]
        newCount = oldCount - 1
        self.myDict[key] = newCount

        self.countDict[oldCount].remove(key)
        if not self.countDict[oldCount]:
            del self.countDict[oldCount]

        if newCount == 0:
            del self.myDict[key]
        else:
            if newCount not in self.countDict:
                self.countDict[newCount] = set()
            self.countDict[newCount].add(key)

        self._updateMinMax()

    def getMaxKey(self) -> str:
        if not self.myDict:
            return ""
        maxKeys = self.countDict.get(self.maxVal, set())
        return next(iter(maxKeys)) if maxKeys else ""

    def getMinKey(self) -> str:
        if not self.myDict:
            return ""
        minKeys = self.countDict.get(self.minVal, set())
        return next(iter(minKeys)) if minKeys else ""
